backend/api/Controllers/EmergencyActionController.cs

```diff
@@ -0,0 +1,167 @@
+                return NotFound("Robot not found");
+            }
+
+            var area = await _areaService.ReadByInstallationAndName(installationCode, areaName);
+            if (area == null)
+            {
+                _logger.LogError("Could not find area {AreaName} for installation code {InstallationCode}", areaName, installationCode);
+                return NotFound("Area not found");
+            }
+
+            _emergencyActionService.TriggerEmergencyButtonPressedForRobot(new EmergencyButtonPressedForRobotEventArgs(robot.Id, area.Id));
+
+            return Ok("Request to abort current mission and move robot back to safe position received");
+        }
+
+        /// <summary>
+        ///     This endpoint will abort the current running mission run and attempt to return the robot to a safe position in the
+        ///     area. The mission run queue for the robot will be frozen and no further missions will run until the emergency
+        ///     action has been reversed.
+        /// </summary>
+        /// <remarks>
+        ///     <para> The endpoint fires an event which is then processed to stop the robot and schedule the next mission </para>
+        /// </remarks>
+        [HttpPost]
+        [Route("{installationCode}/abort-current-missions-and-send-all-robots-to-safe-zone")]
+        [Authorize(Roles = Role.User)]
+        [ProducesResponseType(typeof(MissionRun), StatusCodes.Status200OK)]
+        [ProducesResponseType(StatusCodes.Status400BadRequest)]
+        [ProducesResponseType(StatusCodes.Status401Unauthorized)]
+        [ProducesResponseType(StatusCodes.Status403Forbidden)]
+        [ProducesResponseType(StatusCodes.Status404NotFound)]
+        [ProducesResponseType(StatusCodes.Status500InternalServerError)]
+        public ActionResult AbortCurrentMissionAndSendAllRobotsToSafeZone(
+            [FromRoute] string installationCode)
+        {
+
+            var robots = _robotService.ReadAll().Result.ToList().FindAll(a =>
+                            a.CurrentInstallation.ToLower(CultureInfo.CurrentCulture).Equals(installationCode.ToLower(CultureInfo.CurrentCulture), StringComparison.Ordinal) &&
+                            a.CurrentArea != null);
+
+            foreach (var robot in robots)
+            {
+                _emergencyActionService.TriggerEmergencyButtonPressedForRobot(new EmergencyButtonPressedForRobotEventArgs(robot.Id, robot.CurrentArea!.Id));
+
+            }
+
+            return Ok("Request to abort current mission and move all robots back to safe position received");
+        }
+
+        /// <summary>
+        ///     This query will clear the emergency state that is introduced by aborting the current mission and returning to a
+        ///     safe zone. Clearing the emergency state means that mission runs that may be in the robots queue will start."
+        /// </summary>
+        [HttpPost]
+        [Route("{robotId}/{installationCode}/{areaName}/clear-robot-emergency-state")]
```

Comment:
Do you need to supply the areaName and the installationCode in addition to the robotId? Could we use the area field from the robot object instead?

Reply:
At the moment we are not using this endpoint. This endpoint sends only clears one robot from the safe zone. We are using the one below, where all robots are cleared from the safe zone and that one only gets the installation code to clear all robots in that installation. I kept the code because it might be useful in the future if we want to only clear one area or one deck for example, but can also remove it if you think that is better to avoid confusion